linux/src/threadpool_linux.c

```diff
@@ -0,0 +1,513 @@
+// Copyright (C) Microsoft Corporation. All rights reserved.
+
+#include <stdlib.h>
+#include <inttypes.h>
+#include <stdio.h>
+#include <stdbool.h>
+#include <semaphore.h>
+#include <time.h>
+#include <string.h>
+
+#include "macro_utils/macro_utils.h"
+
+#include "c_logging/xlogging.h"
+
+#include "c_pal/gballoc_hl.h" // IWYU pragma: keep
+#include "c_pal/gballoc_hl_redirect.h"
+#include "c_pal/threadapi.h"
+#include "c_pal/interlocked.h"
+#include "c_pal/sync.h"
+#include "c_pal/srw_lock.h"
+#include "c_pal/execution_engine.h"
+#include "c_pal/execution_engine_linux.h"
+
+#include "c_pal/threadpool.h"
+
+#define TIMEOUT_MS          100
+#define MAX_THREAD_COUNT    1
+#define THREAD_LIMIT        65536
+
+#define TASK_RESULT_VALUES  \
+    TASK_NOT_USED,          \
+    TASK_INITIALIZING,      \
+    TASK_WAITING,           \
+    TASK_WORKING
+
+MU_DEFINE_ENUM(TASK_RESULT, TASK_RESULT_VALUES);
+MU_DEFINE_ENUM_STRINGS(TASK_RESULT, TASK_RESULT_VALUES)
+
+#define THREADPOOL_STATE_VALUES \
+    THREADPOOL_STATE_NOT_OPEN,  \
+    THREADPOOL_STATE_OPENING,   \
+    THREADPOOL_STATE_OPEN,      \
+    THREADPOOL_STATE_CLOSING
+
+MU_DEFINE_ENUM(THREADPOOL_STATE, THREADPOOL_STATE_VALUES)
+MU_DEFINE_ENUM_STRINGS(THREADPOOL_STATE, THREADPOOL_STATE_VALUES)
+
+typedef struct THREADPOOL_TASK_TAG
+{
+    volatile_atomic int32_t task_state;
+    THREADPOOL_WORK_FUNCTION task_func;
+    void* task_param;
+} THREADPOOL_TASK;
+
+typedef struct THREADPOOL_TAG
+{
+    volatile_atomic int32_t state;
+    volatile_atomic int32_t thread_count;
+    uint32_t max_thread_count;
+    uint32_t min_thread_count;
+    int32_t used_thread_count;
+    volatile_atomic int32_t task_count;
+
+    sem_t semaphore;
+
+    // Pool Array
+    THREADPOOL_TASK* task_array;
+    volatile_atomic int64_t insert_idx;
+    volatile_atomic int64_t consume_idx;
+    volatile_atomic int32_t task_array_size;
+    volatile_atomic int32_t pending_call_count;
+
+    SRW_LOCK_HANDLE srw_lock;
+    uint32_t list_index;
+
+    THREAD_HANDLE* thread_handle_array;
+} THREADPOOL;
+
+static void internal_close(THREADPOOL_HANDLE threadpool)
+{
+    bool should_wait_for_transition = false;
+    do
+    {
+        THREADPOOL_STATE current_state = interlocked_compare_exchange(&threadpool->state, THREADPOOL_STATE_CLOSING, THREADPOOL_STATE_OPEN);
+        if (current_state == THREADPOOL_STATE_NOT_OPEN || current_state == THREADPOOL_STATE_CLOSING)
```

Comment:
>THREADPOOL_STATE_CLOSING [](http://example.com/codeflow?start=75&length=24)

I think if we get closing, we should continue waiting to make the close block.
Otherwise the caller of the close would assume it is closed, but it is not yet.